integrations/firehydrant/main.py

```diff
@@ -0,0 +1,157 @@
+from typing import Any
+from enum import StrEnum
+from loguru import logger
+from port_ocean.core.ocean_types import ASYNC_GENERATOR_RESYNC_TYPE
+from port_ocean.context.ocean import ocean
+import asyncio
+from client import FirehydrantClient
+
+
+class ObjectKind(StrEnum):
+    ENVIRONMENT = "environment"
+    INCIDENT = "incident"
+    SERVICE = "service"
+    RETROSPECTIVE = "retrospective"
+
+
+## Helper function to initialize the Firehydrant client
+def init_client() -> FirehydrantClient:
+    return FirehydrantClient(
+        ocean.integration_config["api_url"],
+        ocean.integration_config["token"],
+        ocean.integration_config.get("app_host", ""),
+    )
+
+
+## Enriches the incident report data
+async def enrich_retrospective_with_incident_data(
+    http_client: FirehydrantClient, semaphore: asyncio.Semaphore, report: dict[str, Any]
+) -> dict[str, Any]:
+    async with semaphore:
+        tasks = await http_client.get_tasks_by_incident(report["incident"]["id"])
+        incident_info = {"tasks": tasks}
+        report["__incident"] = incident_info
+        return report
+
+
+## Enriches the services data with incident milestones (used to compute service analytics in jq)
+async def enrich_service_with_incident_data(
+    http_client: FirehydrantClient,
+    semaphore: asyncio.Semaphore,
+    service: dict[str, Any],
+) -> dict[str, Any]:
+    async with semaphore:
+        milestones = await http_client.get_milestones_by_incident(
+            service["active_incidents"]
+        )
```

Comment:
You can use the incidents paginated request to get all the incident that is connected to one of the services in a list

https://developers.firehydrant.com/#/operations/getV1Incidents

![Uploading image.png…]()